enterprise/app/trends/heatmap.tsx

```diff
@@ -0,0 +1,434 @@
+  top: 1,
+  right: 1,
+  bottom: 39,
+  left: 89,
+};
+
+// This is a magic number that states there will only be one axis label for
+// every N pixels of rendered axis length (currently 100).
+const TICK_LABEL_SPACING_MAGIC_NUMBER = 100;
+// This is a discretized dump of the 'Purples' scale from d3-scale-chromatic,
+// starting arbitrarily from .3 because it looked nice.  The scale is great but
+// the package itself is heavy.  The d3-scale-chromatic scale is interpolated
+// from https://colorbrewer2.org/#type=sequential&scheme=Purples&n=3
+const heatmapColorScale = [
+  "#d9d8ea", // .30
+  "#cecee5", // .35
+  "#c2c2df", // ...
+  "#b6b5d8",
+  "#aaa8d0",
+  "#9e9bc9",
+  "#928ec3",
+  "#8782bc",
+  "#7c73b5",
+  "#7363ac",
+  "#6a51a3",
+  "#61409b",
+  "#583093",
+  "#501f8c",
+  "#471084", // 1
+];
+
+type SelectionData = {
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+  selectionXStart: number;
+  selectionYStart: number;
+  selectionXEnd: number;
+  selectionYEnd: number;
+};
+
+class HeatmapComponentInternal extends React.Component<HeatmapProps, State> {
+  state: State = {};
+  svgRef: React.RefObject<SVGSVGElement> = React.createRef();
+  xScaleBand: ScaleBand<number> = scaleBand();
+  yScaleBand: ScaleBand<number> = scaleBand();
+
+  // These ordered pairs indicate which cell was first clicked on by the user.
+  // By keeping them in order, we ensure that as the user moves their mouse, the
+  // selection will "pivot" around that cell.
+  pendingClick?: [SelectedCellData, SelectedCellData];
+  selectedData?: [SelectedCellData, SelectedCellData];
+
+  renderYBucketValue(v: number) {
+    return this.props.metricBucketFormatter(v);
+  }
+
+  private numHeatmapRows(): number {
+    return Math.max(this.props.heatmapData.bucketBracket.length - 1, 1);
+  }
+
+  private computeBucket(x: number, y: number): SelectedCellData | undefined {
+    if (!this.svgRef.current) {
+      return undefined;
+    }
+
+    const { top, left } = this.svgRef.current.getBoundingClientRect();
+    const mouseX = x - left - CHART_MARGINS.left - this.xScaleBand.step() / 2;
+    const mouseY = y - top - CHART_MARGINS.top - this.yScaleBand.step() / 2;
+    const stepX = Math.round(mouseX / this.xScaleBand.step());
+    const stepY = this.yScaleBand.domain().length - Math.round(mouseY / this.yScaleBand.step()) - 1;
+
+    const timestamp = this.props.heatmapData.timestampBracket[stepX];
+    const metric = this.props.heatmapData.bucketBracket[stepY];
+    if (!timestamp || !metric) {
+      return undefined;
+    }
+    const column = this.props.heatmapData.column[stepX];
+    if (!column || column.value.length <= stepY) {
+      return undefined;
+    }
+
+    return {
+      timestamp: +timestamp,
+      timestampBucketIndex: stepX,
+      metric: +metric,
+      metricBucketIndex: stepY,
+      value: +column.value[stepY],
+    };
+  }
+
+  renderTooltip(c?: MouseCoords) {
+    if (!c || this.pendingClick) {
+      return null;
+    }
```

Comment:
nit, IIUC this would work now?

```suggestion
  renderTooltip(c: MouseCoords) {
    if (this.pendingClick) {
      return null;
    }
```

Reply:
Oops, yup, fixed.